x/evm/keeper/bank_extension.go

```diff
@@ -0,0 +1,166 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	auth "github.com/cosmos/cosmos-sdk/x/auth/types"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+
+	"github.com/NibiruChain/nibiru/v2/eth"
+	"github.com/NibiruChain/nibiru/v2/x/evm"
+	"github.com/NibiruChain/nibiru/v2/x/evm/statedb"
+)
+
+var (
+	_ bankkeeper.Keeper     = &NibiruBankKeeper{}
+	_ bankkeeper.SendKeeper = &NibiruBankKeeper{}
+)
+
+type NibiruBankKeeper struct {
+	bankkeeper.BaseKeeper
+	StateDB *statedb.StateDB
+}
+
+func (evmKeeper *Keeper) NewStateDB(
+	ctx sdk.Context, txConfig statedb.TxConfig,
+) *statedb.StateDB {
+	stateDB := statedb.New(ctx, evmKeeper, txConfig)
+	evmKeeper.Bank.ResetStateDB(stateDB)
+	return stateDB
+}
+
+func (bk NibiruBankKeeper) MintCoins(
+	ctx sdk.Context,
+	moduleName string,
+	coins sdk.Coins,
+) error {
+	// Use the embedded function from [bankkeeper.Keeper]
+	if err := bk.BaseKeeper.MintCoins(ctx, moduleName, coins); err != nil {
+		return err
+	}
+	if findEtherBalanceChangeFromCoins(coins) {
+		moduleBech32Addr := auth.NewModuleAddress(evm.ModuleName)
+		bk.SyncStateDBWithAccount(ctx, moduleBech32Addr)
+	}
+	return nil
+}
+
+func (bk *NibiruBankKeeper) ResetStateDB(db *statedb.StateDB) {
+	bk.StateDB = db
+}
+
+// s.Require().Equal(
+// 	statedb.FromVM(evmObj).GetBalance(
+// 		eth.NibiruAddrToEthAddr(randomAcc),
+// 	).String(),
+// 	"420"+strings.Repeat("0", 12),
+// )
+
+func (bk NibiruBankKeeper) BurnCoins(
+	ctx sdk.Context,
+	moduleName string,
+	coins sdk.Coins,
+) error {
+	// Use the embedded function from [bankkeeper.Keeper]
+	if err := bk.BaseKeeper.BurnCoins(ctx, moduleName, coins); err != nil {
+		return err
+	}
+	if findEtherBalanceChangeFromCoins(coins) {
+		moduleBech32Addr := auth.NewModuleAddress(evm.ModuleName)
+		bk.SyncStateDBWithAccount(ctx, moduleBech32Addr)
+	}
+	return nil
+}
+
+func (bk NibiruBankKeeper) SendCoins(
+	ctx sdk.Context,
+	fromAddr sdk.AccAddress,
+	toAddr sdk.AccAddress,
+	coins sdk.Coins,
+) error {
+	// Use the embedded function from [bankkeeper.Keeper]
+	if err := bk.BaseKeeper.SendCoins(ctx, fromAddr, toAddr, coins); err != nil {
+		return err
+	}
+	if findEtherBalanceChangeFromCoins(coins) {
+		bk.SyncStateDBWithAccount(ctx, fromAddr)
+		bk.SyncStateDBWithAccount(ctx, toAddr)
+	}
+	return nil
+}
+
+func (bk *NibiruBankKeeper) SyncStateDBWithAccount(
+	ctx sdk.Context, acc sdk.AccAddress,
+) {
+	// If there's no StateDB set, it means we're not in an EthereumTx.
+	if bk.StateDB == nil {
+		return
+	}
+	balanceWei := evm.NativeToWei(
+		bk.GetBalance(ctx, acc, evm.EVMBankDenom).Amount.BigInt(),
+	)
+	bk.StateDB.SetBalanceWei(eth.NibiruAddrToEthAddr(acc), balanceWei)
+}
+
+func findEtherBalanceChangeFromCoins(coins sdk.Coins) (found bool) {
+	for _, c := range coins {
+		if c.Denom == evm.EVMBankDenom {
+			return true
+		}
+	}
+	return false
```

Comment:
I would agree with that but looked into the assumptions in the method and found them unsafe. I'll walk through that here.

cosmos-sdk@v0.47.11/types/coin.go:
```go
// AmountOf returns the amount of a denom from coins
func (coins Coins) AmountOf(denom string) Int {
	mustValidateDenom(denom)
	return coins.AmountOfNoDenomValidation(denom)
}

// AmountOfNoDenomValidation returns the amount of a denom from coins
// without validating the denomination.
func (coins Coins) AmountOfNoDenomValidation(denom string) Int {
	if ok, c := coins.Find(denom); ok {
		return c.Amount
	}
	return ZeroInt()
}
```

This seems harmless at first glance until you read what the `Coins.Find` function says. `Coin.Find` assumes that the coins are sorted and does a binary search to go through the slice and find coin with the denom specified. That's problematic because it assumes   the sequence of coins is sorted, which can cause strange behavior and panics inside the binary search.

We could forcibly validate the `Coins` argument to make sure it's sorted using `Coins.Sorted` or check the validation error value from `Coins.ValidateBasic`, but both of these activities would end up iterating through the entire slice anyway. Thus, I'd argue it's better in terms of performance and safety to simply iterate through the slice to find the desired denom rather than using binary search or anything that could panic.

Put simply, between these options:
1. basic for loop to find denom  
2. `Coins.Sort` -> `Coins.Find` binary search for denom  
3. `if err := Coins.ValidateBasic(); err == nil` -> `Coins.Find`  
  else do (2) 

I think option (1) is the best. 

```go
// Find returns true and coin if the denom exists in coins. Otherwise it returns false
// and a zero coin. Uses binary search.
// CONTRACT: coins must be valid (sorted).
func (coins Coins) Find(denom string) (bool, Coin) {
	switch len(coins) {
	case 0:
		return false, Coin{}

	case 1:
		coin := coins[0]
		if coin.Denom == denom {
			return true, coin
		}
		return false, Coin{}

	default:
		midIdx := len(coins) / 2 // 2:1, 3:1, 4:2
		coin := coins[midIdx]
		switch {
		case denom < coin.Denom:
			return coins[:midIdx].Find(denom)
		case denom == coin.Denom:
			return true, coin
		default:
			return coins[midIdx+1:].Find(denom)
		}
	}
}
```